File: clevertap-core/src/main/java/com/clevertap/android/sdk/customviews/MediaPlayerRecyclerView.java
package com.clevertap.android.sdk.customviews;
import android.content.Context;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.View;
import android.widget.AbsListView;
import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;
import androidx.annotation.RestrictTo.Scope;
import androidx.core.content.res.ResourcesCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.clevertap.android.sdk.R;
import com.clevertap.android.sdk.inbox.CTInboxBaseMessageViewHolder;

import com.clevertap.android.sdk.video.inbox.ExoplayerHandle;

import kotlin.jvm.functions.Function0;
import kotlin.jvm.functions.Function3;

@RestrictTo(Scope.LIBRARY)
public class MediaPlayerRecyclerView extends RecyclerView {

    private CTInboxBaseMessageViewHolder playingHolder;

    private final ExoplayerHandle handle = new ExoplayerHandle();

    private final Rect rect = new Rect();

    private final OnScrollListener onScrollListener = new OnScrollListener() {
        @Override
        public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
            super.onScrollStateChanged(recyclerView, newState);
            if (newState == AbsListView.OnScrollListener.SCROLL_STATE_IDLE) {
                playVideo();
            }
        }

        @Override
        public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
            super.onScrolled(recyclerView, dx, dy);
        }
    };
    private final OnChildAttachStateChangeListener onChildAttachStateChangeListener = new OnChildAttachStateChangeListener() {
        @Override
        public void onChildViewAttachedToWindow(@NonNull View view) {
        }

        @Override
        public void onChildViewDetachedFromWindow(@NonNull View view) {
            if (playingHolder != null && playingHolder.itemView.equals(view)) {
                stop();
            }
        }
    };;

    /**
     * {@inheritDoc}
     */
    public MediaPlayerRecyclerView(Context context) {
        super(context);
        initialize();
    }

    /**
     * {@inheritDoc}
     */
    public MediaPlayerRecyclerView(
            Context context,
            AttributeSet attrs
    ) {
        super(context, attrs);
        initialize();
    }

    /**
     * {@inheritDoc}
     */
    public MediaPlayerRecyclerView(
            Context context,
            AttributeSet attrs,
            int defStyleAttr
    ) {
        super(context, attrs, defStyleAttr);
        initialize();
    }

    public void onPausePlayer() {
        handle.setPlayWhenReady(false);
    }

    public void onRestartPlayer() {
        initialize();
        playVideo();
    }

    public void playVideo() {
        CTInboxBaseMessageViewHolder targetHolder = findBestVisibleMediaHolder();

        // Case 1 : No viewholder has video item in it
        if (targetHolder == null) {
            removeVideoView();
            return;
        }

        // Case 2 : Found viewholder is same with surface and player attached
        if (playingHolder != null && playingHolder.itemView.equals(targetHolder.itemView)) {
            boolean measured = playingHolder.itemView.getGlobalVisibleRect(rect);
            int visibleHeight = measured ? rect.height() : 0;
            boolean play = visibleHeight >= 400;
            if (play && playingHolder.shouldAutoPlay()) {
                handle.setPlayWhenReady(true);
            } else {
                handle.setPlayWhenReady(false);
            }
            return;
        }

        // Case 3: Video has to be played in different view holder so we remove and reattch to correct one
        removeVideoView();
        float currentVolume = handle.playerVolume();
        boolean addedVideo = targetHolder.addMediaPlayer(
                currentVolume,
                new Function0<Float>() {
                    @Override
                    public Float invoke() {
                        handle.handleMute();
                        return handle.playerVolume();
                    }
                },
                new Function3<String, Boolean, Boolean, Void>() {
                    @Override
                    public Void invoke(String uri, Boolean isMediaAudio, Boolean isMediaVideo) {
                        handle.startPlaying(
                                getContext(),
                                uri,
                                isMediaAudio,
                                isMediaVideo
                        );
                        return null;
                    }
                },
                handle.player()
        );
        handle.playMedia();
        if (addedVideo) {
            playingHolder = targetHolder;
        }
    }

    public void stop() {
        /*if (player != null) {
            player.stop();
        }*/
        handle.pause();
        playingHolder = null;
    }

    private CTInboxBaseMessageViewHolder findBestVisibleMediaHolder() {
        CTInboxBaseMessageViewHolder bestHolder = null;

        //noinspection ConstantConditions
        int startPosition = ((LinearLayoutManager) getLayoutManager()).findFirstVisibleItemPosition();
        int endPosition = ((LinearLayoutManager) getLayoutManager()).findLastVisibleItemPosition();

        int bestHeight = 0;
        for (int i = startPosition; i <= endPosition; i++) {
            int pos = i - startPosition;
            View child = getChildAt(pos);
            if (child == null) {
                continue;
            }
            CTInboxBaseMessageViewHolder holder = (CTInboxBaseMessageViewHolder) child.getTag();
            if (holder != null) {
                if (!holder.needsMediaPlayer()) {
                    continue;
                }
                boolean measured = holder.itemView.getGlobalVisibleRect(rect);
                int height = measured ? rect.height() : 0;
                if (height > bestHeight) {
                    bestHeight = height;
                    bestHolder = holder;
                }
            }
        }
        return bestHolder;
    }

    private void initialize() {

        handle.initExoplayer(getContext().getApplicationContext(), bufferingStarted(), playerReady());
        handle.initPlayerView(getContext().getApplicationContext(), artworkAsset());

        recyclerViewListeners();
    }

    private Function0<Void> bufferingStarted() {
        return new Function0<Void>() {
            @Override
            public Void invoke() {
                if (playingHolder != null) {
                    playingHolder.playerBuffering();
                }
                return null;
            }
        };
    }

    private Function0<Void> playerReady() {
        return new Function0<Void>() {
            @Override
            public Void invoke() {
                if (playingHolder != null) {
                    playingHolder.playerReady();
                }
                return null;
            }
        };
    }

    private Function0<Drawable> artworkAsset() {
        return new Function0<Drawable>() {
            @Override
            public Drawable invoke() {
                return ResourcesCompat.getDrawable(getResources(), R.drawable.ct_audio, null);
            }
        };
    }

    private void recyclerViewListeners() {
        removeOnScrollListener(onScrollListener);
        removeOnChildAttachStateChangeListener(onChildAttachStateChangeListener);
        addOnScrollListener(onScrollListener);
        addOnChildAttachStateChangeListener(onChildAttachStateChangeListener);
    }

    private void removeVideoView() {
        handle.pause();
        if (playingHolder != null) {
            playingHolder.playerRemoved(); // removes all the views from video container
            playingHolder = null;
        }
    }
}

File: clevertap-core/src/main/java/com/clevertap/android/sdk/customviews/MediaPlayerRecyclerView.kt
package com.clevertap.android.sdk.customviews

import android.content.Context
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.View
import android.widget.AbsListView
import androidx.annotation.RestrictTo
import androidx.core.content.res.ResourcesCompat
import androidx.media3.common.util.UnstableApi
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.clevertap.android.sdk.R
import com.clevertap.android.sdk.inbox.CTInboxBaseMessageViewHolder
import com.clevertap.android.sdk.video.InboxVideoPlayerHandle
import com.clevertap.android.sdk.video.VideoLibChecker
import com.clevertap.android.sdk.video.VideoLibraryIntegrated
import com.clevertap.android.sdk.video.inbox.ExoplayerHandle
import com.clevertap.android.sdk.video.inbox.Media3Handle

@UnstableApi
@RestrictTo(RestrictTo.Scope.LIBRARY)
class MediaPlayerRecyclerView : RecyclerView {

    private val handle : InboxVideoPlayerHandle = when (VideoLibChecker.mediaLibType) {
        VideoLibraryIntegrated.MEDIA3 -> {
            Media3Handle()
        }
        else -> {
            ExoplayerHandle()
        }
    }
    private val rect = Rect()
    private val onScrollListener: OnScrollListener = object : OnScrollListener() {
        override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
            super.onScrollStateChanged(recyclerView, newState)
            if (newState == AbsListView.OnScrollListener.SCROLL_STATE_IDLE) {
                playVideo()
            }
        }
    }
    private val onChildAttachStateChangeListener: OnChildAttachStateChangeListener =
        object : OnChildAttachStateChangeListener {
            override fun onChildViewAttachedToWindow(view: View) {}
            override fun onChildViewDetachedFromWindow(view: View) {
                playingHolder?.let { ph ->
                    if (ph.itemView == view) {
                        stop()
                    }
                }
            }
        }

    private var playingHolder: CTInboxBaseMessageViewHolder? = null

    /**
     * {@inheritDoc}
     */
    constructor(context: Context) : super(context) {
        initialize()
    }

    /**
     * {@inheritDoc}
     */
    constructor(
        context: Context,
        attrs: AttributeSet
    ) : super(context, attrs) {
        initialize()
    }

    /**
     * {@inheritDoc}
     */
    constructor(
        context: Context,
        attrs: AttributeSet,
        defStyleAttr: Int
    ) : super(context, attrs, defStyleAttr) {
        initialize()
    }

    fun onPausePlayer() {
        handle.setPlayWhenReady(false)
    }

    fun onRestartPlayer() {
        initialize()
        playVideo()
    }

    fun playVideo() {
        val targetHolder = findBestVisibleMediaHolder()

        // Case 1 : No viewholder has video item in it
        if (targetHolder == null) {
            removeVideoView()
            return
        }

        // Case 2 : Found viewholder is same with surface and player attached
        playingHolder?.let { ph ->
            if (ph.itemView == targetHolder.itemView) {
                val measured = ph.itemView.getGlobalVisibleRect(rect)
                val visibleHeight = if (measured) {
                    rect.height()
                } else {
                    0
                }
                val play = visibleHeight >= 400
                if (play && ph.shouldAutoPlay()) {
                    handle.setPlayWhenReady(true)
                } else {
                    handle.setPlayWhenReady(false)
                }
                return
            } else {
                // no-op
            }
        }

        // Case 3: Video has to be played in different view holder so we remove and reattch to correct one
        removeVideoView()
        initialize()
        val currentVolume = handle.playerVolume()
        val addedVideo = targetHolder.addMediaPlayer(
            currentVolume,
            {
                handle.handleMute()
                handle.playerVolume()
            },
            { uri: String, isMediaAudio: Boolean, isMediaVideo: Boolean ->
                handle.startPlaying(
                    ctx = context,
                    uriString = uri,
                    isMediaAudio = isMediaAudio,
                    isMediaVideo = isMediaVideo
                )
                null
            },
            handle.videoSurface()
        )
        if (addedVideo) {
            playingHolder = targetHolder
        }
    }

    fun stop() {
        /*if (player != null) {
            player.stop();
        }*/
        handle.pause()
        playingHolder = null
    }

    private fun findBestVisibleMediaHolder(): CTInboxBaseMessageViewHolder? {
        var bestHolder: CTInboxBaseMessageViewHolder? = null
        val startPosition = (layoutManager as LinearLayoutManager?)?.findFirstVisibleItemPosition() ?: 0
        val endPosition = (layoutManager as LinearLayoutManager?)?.findLastVisibleItemPosition() ?: 0
        var bestHeight = 0
        for (i in startPosition..endPosition) {
            val pos = i - startPosition
            val child = getChildAt(pos) ?: continue

            val holder = child.tag as? CTInboxBaseMessageViewHolder
            if (holder != null) {
                if (!holder.needsMediaPlayer()) {
                    continue
                }
                val measured = holder.itemView.getGlobalVisibleRect(rect)
                val height = if (measured) {
                    rect.height()
                } else {
                    0
                }
                if (height > bestHeight) {
                    bestHeight = height
                    bestHolder = holder
                }
            }
        }
        return bestHolder
    }

    private fun initialize() {
        handle.initExoplayer(
            context = context.applicationContext,
            buffering = ::bufferingStarted,
            playerReady = ::playerReady
        )
        handle.initPlayerView(
            context = context.applicationContext,
            artworkAsset = ::artworkAsset
        )
        recyclerViewListeners()
    }

    private fun bufferingStarted() {
        playingHolder?.playerBuffering()
    }

    private fun playerReady() {
        playingHolder?.playerReady()
    }

    private fun artworkAsset(): Drawable {
        return ResourcesCompat.getDrawable(resources, R.drawable.ct_audio, null)!!
    }

    private fun recyclerViewListeners() {
        removeOnScrollListener(onScrollListener)
        removeOnChildAttachStateChangeListener(onChildAttachStateChangeListener)
        addOnScrollListener(onScrollListener)
        addOnChildAttachStateChangeListener(onChildAttachStateChangeListener)
    }

    private fun removeVideoView() {
        handle.pause()
        playingHolder?.playerRemoved() // removes all the views from video container
    }
}
